Teuchos Numerics API Refactor

krcb opened the pull request:
This PR is a work in progress to address issue #1208. Initially, this addresses the Teuchos::LAPACK interface.

krcb commented:
@bartlettroscoe, @jwillenbring this is the PR that addresses the effort that we discussed this morning.

jwillenbring commented:
@maherou asked about this today so I wanted to mention him on the PR

bartlettroscoe commented on packages/teuchos/numerics/src/Teuchos_LAPACK.cpp:
```diff
@@ -64,7 +64,7 @@ functions that include the macro:
 namespace {
 
 #if defined (INTEL_CXML)
-        unsigned int& one=1;
```
Did the compiler or tests on 'ride' catch this problem?

bartlettroscoe commented on packages/teuchos/numerics/src/Teuchos_LAPACK.cpp:
```diff
@@ -1236,11 +1235,11 @@ namespace Teuchos
 
   void LAPACK<int, std::complex<float> >::GEES(const char& JOBVS, const int& n, std::complex<float>* A, const int& lda, int* sdim, float* WR, float* WI, std::complex<float>* VS, const int& ldvs, std::complex<float>* WORK, const int& lwork, float* RWORK, int* BWORK, int* info) const
   {
-    int& (*nullfptr)(std::complex<float>*) = NULL;
```
Did compiler or tests on 'ride' catch this?

bartlettroscoe commented on packages/teuchos/numerics/src/Teuchos_LAPACK.cpp:
```diff
@@ -1236,11 +1235,11 @@ namespace Teuchos
-    const char& sort = 'N';
+    const char sort = 'N';
     CGEES_F77(CHAR_MACRO(JOBVS), CHAR_MACRO(sort), nullfptr, &n, A, &lda, sdim, &W[0], VS, &ldvs, WORK, &lwork, RWORK, BWORK, info);
-    for (int& i=0; i<n; i++) {
```
Can't believe that even compiled before this change.

krcb commented:
> I am curious about some of the other changes that removed & from some local variables. How did you catch those? Did those generate compiler warnings or errors or failing tests on 'ride' or any other machine?

These were created by the script I used to perform the refactor being too aggressive in adding '&'; my first few reviews of the code failed to catch these. These are now fixed by subsequent commits, so the only changes in the final state of this PR should be in the API. Assigning to @MicheldeMessieres to test.

bartlettroscoe commented:
> These were created by the script I used to perform the refactor being too aggressive in adding '&'; my first few reviews of the code failed to catch these. These are now fixed by subsequent commits, so the only changes in the final state of this PR should be in the API.

Okay, just make sure and squash these commits together with `rebase -i @{u}` before you push.

krcb commented:
@MicheldeMessieres appears to have discovered a bug in the Teuchos::LAPACK API that the modifications on this PR have exposed. Specifically, in the Teuchos::LAPACK API, we have:

```
void TREXC(const char COMPQ, const OrdinalType n, ScalarType* T, const OrdinalType ldt, ScalarType* Q, const OrdinalType ldq, OrdinalType ifst, OrdinalType ilst, ScalarType* WORK, OrdinalType* info) const;
```

which calls (for the double specialization):

```
DTREXC_F77(Teuchos_fcd, const int* n, double* t, const int* ldt, double* q, const int* ldq, int* ifst, int* ilst, double* work, int* info);
```

Note that ifst and ilst are not declared const in the DTREXC_F77 interface. This is because these variables are modified in the DTREXC routine. However, the CXX interface has them passed by value, so these are not returned to the user (which is inconsistent with the LAPACK interface, at least at v3.7.0). We have three options:

1. Change the CXX interface to pass by pointer in order to maintain consistency with the LAPACK API. This breaks backward compatibility.

2. Change the CXX interface to pass by const reference, create a non-const copy inside the CXX implementation and pass that into DTREXC_F77 wrapper. This would maintain backwards compatibility, but would be inconsistent with the LAPACK API.

3. Implement both 1) and 2), mark the backwards compatible interface as deprecated and update the usage of this interface in Trillinos to use the new interface.

Which of these is preferable? Or is there an alternative that we have not considered here?

MicheldeMessieres commented:
Searching Teuchos_LAPACK, tried to get a complete list of where the same type of issue comes up:
TREXC (ifst, ilst not passed as int *)
GEBAL (ilo, ihi not passed as int *)
POSIX (EQUED not passed as char *)
GESVX (EQUED not passed as char *)

Did not find any issues like this for for Teuchos_BLAS.

bartlettroscoe commented:
> Note that ifst and ilst are not declared const in the DTREXC_F77 interface. This is because these variables are modified in the DTREXC routine. However, the CXX interface has them passed by value, so these are not returned to the user (which is inconsistent with the LAPACK interface, at least at v3.7.0).

Actually, what I would do is to provide an overload of these functions and pass the variables being modified by pointer (because they are being modified, non-const reference arguments are not great for arguments that are modified) and then deprecate the existing functions that take the pass-by-value copy arguments.  Since these are meant to be direct wrappers for the LAPACK routines, they should have the same semantics for the arguments as they are in Fortran.

Just my opinion.

bartlettroscoe commented:
@MicheldeMessieres AND @krcb,

It seems there is some urgency to get the basic issue with calling LAPACK resolved on 'ride' that is breaking the Belos tests (and therefore some downstream tests as well).

What is the full status of the Trilinos test suite on 'ride' with this current branch?  Are there any failing tests due to the problem with the `ifst` and `ilst` arguments you describe [above](https://github.com/trilinos/Trilinos/pull/1284#issuecomment-299293881)?  If not, then we should merge the current branch minus the commits for the `ifst` and `ilst` arg fix and then open a new issue and/or PR for these new problems and not block #1208.

@rppawlo and @bathmatt, what specific tests do you need to have passing so that this is not a blocker on 'ride'?

krcb commented:
We have got the PR to the state where the Teuchos LAPACK interface is fixed. This resolves one of the Belos test failures on ride.

As mentioned above, we have found a range of inconsistencies in the Teuchos::LAPACK interface that meant downstream usage had to be fixed. We have now also found inconsistencies in the Teuchos LAPACK F77 wrappers compared to how the corresponding LAPACK functions are defined. These have been fixed, but the changes break backwards compatibility and has impact in Trillinos packages. We are still testing to determine the extent of these issues. Finally, Epetra implements it's own LAPACK wrapper (which is used, e.g., by Amesos). This interface likewise needs to be addressed and refactored.

krcb commented:
@MicheldeMessieres has some additional commits that are needed for complex types. We will be pushing these onto this branch momentarily. These resolve failing tests due to the ```ifst``` and ```ilst``` arguments.

rppawlo commented:
I'm not sure about Trilinos test failures, but we have a failure in Drekar for the Anasazi BlockKrylovSchur solver. This solver uses 4 lapack calls, three of which pass in char as the first argument. It was definitely corrupting stack memory in the solver. Once a merge is made, I can test to see if this fixes it.

bartlettroscoe commented:
@MicheldeMessieres AND @krcb,

It seems there is some urgency to get the basic issue with calling LAPACK resolved on 'ride' that is breaking the Belos tests (and therefore some downstream tests as well).

What is the full status of the Trilinos test suite on 'ride' with this current branch?  Are there any failing tests due to the problem with the `ifst` and `ilst` arguments you describe [above](https://github.com/trilinos/Trilinos/pull/1284#issuecomment-299293881)?  If not, then we should merge the current branch minus the commits for the `ifst` and `ilst` arg fix and then open a new issue and/or PR for these new problems and therefore not block #1208 with these separate problems.

@rppawlo and @bathmatt, what specific tests do you need to have passing on 'ride'?

bartlettroscoe commented:
@krcb,

> We have got the PR to the state where the Teuchos LAPACK interface is fixed. This resolves one of the Belos test failures on ride.

Can we chat briefly about the current status of these changes?  It would be good if we could make the minimal changes needed to get Belos to pass.

@rppawlo,

> I'm not sure about Trilinos test failures, but we have a failure in Drekar for the Anasazi BlockKrylovSchur solver. This solver uses 4 lapack calls, three of which pass in char as the first argument. It was definitely corrupting stack memory in the solver.

Can you list the 4 LAPACK calls that are the issue?  Perhaps we can fix just these to start with and merge to 'master' and then address the rest of the issues in a follow-up PR?

rppawlo commented:
The anasazi solver we are using calls these four lapack routines: TREVC, GEES, TREXC, LAPY2

bartlettroscoe commented:
> The anasazi solver we are using calls these four lapack routines: TREVC, GEES, TREXC, LAPY2

Darn, TREXC is the problematic function call with the `ifst` and `ilst` arguments listed [above](https://github.com/trilinos/Trilinos/pull/1284#issuecomment-299293881).

@krcb and @MicheldeMessieres, what is the current ETA for getting the issues with the PR resolved?  Can we get a compact update on where things stand what is left to do?  Just let me know when there is something new to review where I can provide feedback.

Thanks,

-Ross

bartlettroscoe commented:
Just had a good conversation with @krcb and @MicheldeMessieres about the current status of this PR and the problems in general in #1191 on 'ride'.  I will let them provide their summary, but in short, it looks like they are very close to resolving the LAPACK-related issues breaking the Belos tests on 'ride'.  The only downside is that this will technically break backward compatibility for the raw `extern "C"` Fortran wrappers (but those exist just to be called by the templated C++ wrappers which will be maintaining backward compatibility with deprecated functions).

However, @krcb emphasized that these LAPACK calling issues will likely **not** fix all of the failing Belos tests shown in #1191.  There are segfaults in those tests that don't look to be related to the LAPACK calling issues and are likely instead related to the Kokkos/MueLu issues described in #1216.   Therefore, don't expect all of Trilinos problems on 'ride' to be resolved by fixing the calls to LAPACK.

MicheldeMessieres commented:
Updated the PR.

Ran SEMS non-complex and complex using:
./checkin-test-sems.sh --st-extra-builds=MPI_RELEASE_DEBUG_SHARED_PT_COMPLEX -j10 --enable-all-packages=on --local-do-all

All tests passed for both builds.

bartlettroscoe commented:
Has this updated branch been tested on 'ride'?  I think once that is done, we can merge this to 'develop'.  Does everyone agree?

krcb commented:
I have tested that the STEQR test and the affected Belos test passes there (as discussed on #1208). I have not yet had time to run the entire branch there; @MicheldeMessieres would have to get setup there and run the branch.

jwillenbring commented:
@jjellio or @mhoemmen Might it be possible for you to test this PR on ride? That could potentially happen much more quickly than asking @MicheldeMessieres to do so since he has not used the platform in the past. I am just trying to figure out how we can get this done most efficiently.

krcb commented:
Some availability opened up in my schedule, so I'm able to test out Teuchos and Belos this morning on ride. I will update with results shortly.

krcb commented:
For Teuchos Numerics, with ```-DTeuchos_ENABLE_COMPLEX=OFF``` and ```-DTeuchos_ENABLE_COMPLEX=ON```, built with gcc 5.4.0 and OpenBLAS 0.2.19 all tests pass:

```
        Start  94: TeuchosNumerics_BLAS_test
 94/111 Test  #94: TeuchosNumerics_BLAS_test .................................................   Passed    0.15 sec
        Start  95: TeuchosNumerics_BLAS_tmpl_test
 95/111 Test  #95: TeuchosNumerics_BLAS_tmpl_test ............................................   Passed    0.02 sec
        Start  96: TeuchosNumerics_BLAS_tmpl_comp_test
 96/111 Test  #96: TeuchosNumerics_BLAS_tmpl_comp_test .......................................   Passed    0.14 sec
        Start  97: TeuchosNumerics_BLAS_ROTG_test
 97/111 Test  #97: TeuchosNumerics_BLAS_ROTG_test ............................................   Passed    0.02 sec
        Start  98: TeuchosNumerics_DenseSolver_test
 98/111 Test  #98: TeuchosNumerics_DenseSolver_test ..........................................   Passed    0.06 sec
        Start  99: TeuchosNumerics_BandDenseSolver_test
 99/111 Test  #99: TeuchosNumerics_BandDenseSolver_test ......................................   Passed    0.07 sec
        Start 100: TeuchosNumerics_SpdDenseSolver_test
100/111 Test #100: TeuchosNumerics_SpdDenseSolver_test .......................................   Passed    0.04 sec
        Start 101: TeuchosNumerics_QRDenseSolver_test
101/111 Test #101: TeuchosNumerics_QRDenseSolver_test ........................................   Passed    0.02 sec
        Start 102: TeuchosNumerics_LAPACK_test
102/111 Test #102: TeuchosNumerics_LAPACK_test ...............................................   Passed    0.08 sec
        Start 103: TeuchosNumerics_Polynomial_test
103/111 Test #103: TeuchosNumerics_Polynomial_test ...........................................   Passed    0.02 sec
        Start 104: TeuchosNumerics_MatrixMarket_Raw_InOutTest
104/111 Test #104: TeuchosNumerics_MatrixMarket_Raw_InOutTest ................................   Passed    0.03 sec
        Start 105: TeuchosNumerics_BLAS_example
105/111 Test #105: TeuchosNumerics_BLAS_example ..............................................   Passed    0.02 sec
        Start 106: TeuchosNumerics_DenseMatrix_example
106/111 Test #106: TeuchosNumerics_DenseMatrix_example .......................................   Passed    0.02 sec
        Start 107: TeuchosNumerics_SymDenseMatrix_example
107/111 Test #107: TeuchosNumerics_SymDenseMatrix_example ....................................   Passed    0.02 sec
        Start 108: TeuchosNumerics_hilbert
108/111 Test #108: TeuchosNumerics_hilbert ...................................................   Passed    0.02 sec
        Start 109: TeuchosNumerics_LAPACK_example
109/111 Test #109: TeuchosNumerics_LAPACK_example ............................................   Passed    0.03 sec
```

For the rest of Teuchos (Core, ParameterList, Comm), all except one test pass:

```43 - TeuchosCore_TypeConversions_UnitTest (Failed)```

The error reported there appears to be driven by:

```Teuchos::ValueTypeConversionTraits<float, std::string>::convert: The value in the given string "-1.79769313e+308" overflows float.```

krcb commented:
For Belos with this same compiler/BLAS compiler configuration,  the good news is that the Belos test that was failing due to the LAPACK error passes:

```
54/59 Test #54: Belos_Tpetra_PseudoBlockCG_hb_test .............   Passed    0.39 sec
```

The bad news is that the following tests fail:

```
	 10 - Belos_pseudo_gmres_hb (SEGFAULT)
	 11 - Belos_bl_pgmres_hb_0 (SEGFAULT)
	 12 - Belos_bl_pgmres_hb_1 (SEGFAULT)
	 13 - Belos_bl_pgmres_hb_2 (SEGFAULT)
	 15 - Belos_bl_fgmres_hb (SEGFAULT)
	 28 - Belos_gcrodr_hb (SEGFAULT)
	 30 - Belos_prec_gcrodr_hb_1 (SEGFAULT)
```

Running the first of these through gdb seems to indicate an issue in an Epetra_MultiVector constructor:

```
#0  0x00000000101b8d50 in Epetra_MultiVector::Epetra_MultiVector(Epetra_DataAccess, Epetra_MultiVector const&, int*, int) ()
#1  0x00000000100a0808 in Belos::MultiVecTraits<double, Epetra_MultiVector>::CloneView(Epetra_MultiVector const&, std::vector<int, std::allocator<int> > const&) ()
#2  0x00000000100df070 in Belos::BlockGmresIter<double, Epetra_MultiVector, Epetra_Operator>::iterate() ()
#3  0x00000000100e9a08 in Belos::BlockGmresSolMgr<double, Epetra_MultiVector, Epetra_Operator>::solve() ()
#4  0x0000000010078eac in main ()
```

Unfortunately, this suggests that there are further issues to address on this platform for Belos that are not resolved by the changes applied to Teuchos::LAPACK/BLAS in this PR.

bartlettroscoe commented:
> The bad news is that the following tests fail:

@krcb, if those same tests failed before the changes on this branch, I say push this and then we can open a new targeted issue to address those problems (and block #1191 with that new issue).

krcb commented:
@bartlettroscoe The Belos list is consistent with those failures reported originally reported [here](https://github.com/trilinos/Trilinos/issues/1191#issue-217682302), but now with the ```Belos_Tpetra_PseudoBlockCG_hb_test``` passing.

I also built and tested Anasazi to investigate @rppawlo\s reported failure above. While I see many failures, the following test did report as passed:

```
59/72 Test #59: Anasazi_BlockKrylovSchurThyra_test ......................   Passed    4.45 sec
```

This may address the use case described above. I will now check whether the reported failures also fail on develop

krcb commented:
Building a testing current develop for Teuchos and Belos on ride gives a list of failures consistent with that reported above.

krcb commented:
As requested, here is a summary of the changes that this PR implements:

1. In order to address CXX/Fortran interface issues observed on an advanced architecture, the Teuchos::LAPACK and Teuchos::BLAS were refactored to replace pass by value with pass by const reference. The rational behind this is documented [here](https://github.com/trilinos/Trilinos/issues/1208).
2. Then we checked for mismatches with the LAPACK Fortran API. This revealed the following issues:
(a) GESVX and POSVX: EQUED is documented by LAPACK as IN/OUT and can be modified by the Fotran routine. In the current Teuchos::LAPACK API, this parameter is defined as pass by value. In order to make the interface consistent with the LAPACK API, the following changes were made:
    -  In Teuchos_LAPACK_wrappers.hpp, the EQUED char should be char *.
    -  In Teuchos_LAPACK.cpp/hpp the EQUED char should be char *.
    - In  Teuchos_LAPACK_wrappers.hpp, we added a #define Teuchos_nonconst_fcd. This is the char *      equivalent of Teuchos_fcd (which is const char*).That was then used for all the EQUED cases.
    - In Teuchos_LAPACK.cpp/hpp we added a CHARPTR_MACRO. This is the char* equivalent of CHAR_MACRO (which passed &char).That was then applied to new GESVX and POSVX methods using char *EQUED.
    - Deprecated forms were created which take const char& for backwards compatibility. The implementation of the deprecated forms in Teuchos_LAPACK.cpp copies this value to pass to the new correct F77 wrapper by char*, consistent with the original approach.

    (b) TREXC: ifst and ilst is documented by LAPACK as IN/OUT and can be modified by the Fotran routine for real type. For complex types, these variables are documented as IN. In the current Teuchos::LAPACK API, this parameter is defined as pass by value. In order to make the interface consistent with the LAPACK API, the following changes were made:
    -  In Teuchos_LAPACK.cpp/hpp all forms should pass ifst and ilst by int*, not int.
    -  In Teuchos_LAPACK_wrappers.hpp the complex forms should be const int*, not int*. Note that TREXC has a special status since float/double forms modify ifst and ilst, while complex forms do not. Teuchos_LAPACK_wrappers.hpp was changed so the complex forms properly take const int*.
    -  Teuchos_LAPACK.cpp/hpp created new methods which take int* for all types (complex and non-complex). This allows the template specialization to work for all forms.
    -  Deprecated forms were created which take const int& for backwards compatibility. Teuchos_LAPACK.cpp copies this value to pass to the F77 wrappers (for complex, using the new wrapper), consistent with the original setup.     

    (c) GEBAL: ilo and ihi is documented by LAPACK as IN/OUT and can be modified by the Fotran routine. In order to make the interface consistent with the LAPACK API, the following changes were made:
    - In Teuchos_LAPACK.cpp/hpp all forms should pass ilo and ihi by int*, not int.
    - Teuchos_LAPACK.cpp/hpp was changed to pass by int *.
    - We added deprecated forms of GEBAL which take const int& for backwards compatibility.
    - Teuchos_LAPACK.cpp copies this value to pass to the F77 wrapper (which was correct) , consistent with the original setup.

3. Add STEQR test to probe Teuchos::LAPACK char usage in the TeuchosNumerics LAPACK test

4. Modified anasazi package for consistency with modified API's. Specifically, usage of TREXC was changed to pass to int* instead of by value.This means that the Anasazi usage of TREXC is now using the new form, not the deprecated form. This appears to be the only tested place in Trilinos which calls one of API calls that was changed due to inconsistency with LAPACK.

bartlettroscoe commented:
I will now pull and merge this branch to my local 'develop' branch then run the checkin-test-sems.sh script to test and push.  This will ensure that the integrated develop branch 100% passes the CI build.  I will close this PR manually once that is complete.

bartlettroscoe commented:
I just pushed the merge commit:

```
commit 10547a3f7628d055d0ecee56e9d16807ae0d74f6
Merge: 353a943 82d957a
Author: Roscoe A. Bartlett <rabartl@sandia.gov>
Date:   Wed May 10 18:40:13 2017 -0600

    Merge remote-tracking branch 'tech-x-corp/teuchosLAPACKRefactor' into develop
    
    Should address #1208
    
    Build/Test Cases Summary
    Enabled Packages: Anasazi, TeuchosNumerics
    Disabled Packages: PyTrilinos,Claps,TriKota
    Enabled all Forward Packages
    0) MPI_RELEASE_DEBUG_SHARED_PT => passed: passed=2251,notpassed=0 (198.65 min)
    Other local commits for this build/test group: 82d957a, 60eca48, ce20596, 076a26f
```

The other advantage of using the checkin-test-sems.sh script over the 'Merge' button, is that it marks known tested commits and that can be used in a robust git bisection process as described at:

* https://tribits.org/doc/TribitsDevelopersGuide.html#using-git-bisect-with-checkin-test-py-workflows
